### ISSUE TITLE
how to create my own dataset without in memory

### ISSUE THREAD

**Issue by mrqrs** [rusty1s/pytorch_geometric#1473]
## ❓ Questions & Help

    def process(self):
        split = osp.basename(self.root)
        if split == 'train':
            sequences = ['00', '01', '02', '03', '04', '05', '06', '07', '09', '10']
        if split == 'eval':
            sequences = ['08']
        if split == 'test':
            sequences = ['11', '12', '13', '14', '15', '16', '17']
        x_files = get_files_sequence(self.root, sequences, 'voxels', 'npy')
        edge_files = get_files_sequence(self.root, sequences, 'edges', 'npy')
        label_files = get_files_sequence(self.root, sequences, 'labels', 'npy')
        print('processing')
        i = 0
        for j in range(len(x_files)):
            print('Processing:', j, '/', len(x_files))
            voxels_list = np.load(x_files[j], allow_pickle=True)
            edges_list = np.load(edge_files[j], allow_pickle=True)
            labels_list = np.load(label_files[j], allow_pickle=True)
            for m in range(len(voxels_list)):
                x = voxels_list[m]
                num_nodes = x.shape[0]
                edge_index = edges_list[m].T
                label = labels_list[m]
                data = Data(x=x, edge_index=edge_index, y=label, num_nodes=num_nodes)
                torch.save(data, osp.join(self.processed_dir, 'data_{}.pt'.format(i)))
                i += 1
        print('Processed!')

    def len(self):
        files = get_files(self.processed_dir, 'pt')
        return len(files)

    def get(self, idx):
        data = torch.load(osp.join(self.processed_dir, 'data_{}.pt'.format(idx)))
        return data
---------------------------
i used this code to create my own dataset, and it work well. However, when i use Dataloader to loader data, it has some problems.
train_loader = DataLoader(dataset=train_data, batch_size=batch_size)
 data_iter = iter(train_loader)
    for steps in range(int(num_train/batch_size)):
        data = next(data_iter)
next: return a Data list, it has batch_size Data, it can not use to train net with mini-batch samples. How can i do to solve this problem or do i need to write collate function to convert the Data list to a Data? Thank you very much.

**Comment by rusty1s** [comment]
Are you using the PyTorch `DataLoader` or the one provided by PyG? The latter should already take care of this.

**Comment by mrqrs** [comment]
> Are you using the PyTorch `DataLoader` or the one provided by PyG? The latter should already take care of this.

yes, I used pyG dataloader to process data,
In this line:
for data in trainloader:
The return data is a Data list. The x and edge_index are list. It cannot be input in the GCN layer.

Laterally, i merge the Data list manually and it work well.  Is there has merge or collate function in the pyG?

**Comment by rusty1s** [comment]
What do you mean by this? Does it work now? It would be great if you could show me a simple example of your problem, I cannot see any mistake from your code above.

**Comment by mrqrs** [comment]
> What do you mean by this? Does it work now? It would be great if you could show me a simple example of your problem, I cannot see any mistake from your code above.

![image](https://user-images.githubusercontent.com/26370029/88645252-7f083d00-d0f6-11ea-99d1-46e8f17fe8d9.png)
Such as the image, the return data is a Data list, x and edge_index are list. It cannot input to the GCN layer. I use torch.cat to merge x and edge_index separately, so the x and edge_index are tensor. then x, edge_index and data.batch can input to the layer and do batch work.

**Comment by rusty1s** [comment]
I see. That's because you are passing `numpy` tensors to `Data` inside `process`. You should convert them to PyTorch tensors first.

**Comment by mrqrs** [comment]
Thank you very much. Do you mean that i need to convert the x and edge_index to tensors in process before save data?  The process function like this:
` def process(self):
        split = osp.basename(self.root)
        if split == 'train':
            sequences = ['00', '01', '02', '03', '04', '05', '06', '07', '09', '10']
        if split == 'eval':
            sequences = ['08']
        if split == 'test':
            sequences = ['11', '12', '13', '14', '15', '16', '17']
        x_files = get_files_sequence(self.root, sequences, 'voxels', 'npy')
        edge_files = get_files_sequence(self.root, sequences, 'edges', 'npy')
        label_files = get_files_sequence(self.root, sequences, 'labels', 'npy')
        print('processing')
        i = 0
        for j in range(len(x_files)):
            print('Processing:', j, '/', len(x_files))
            voxels_list = np.load(x_files[j], allow_pickle=True)
            edges_list = np.load(edge_files[j], allow_pickle=True)
            labels_list = np.load(label_files[j], allow_pickle=True)
            for m in range(len(voxels_list)):
                x = voxels_list[m]
                num_nodes = x.shape[0]
                edge_index = edges_list[m].T
                label = labels_list[m]
                data = Data(x=x, edge_index=edge_index, y=label, num_nodes=num_nodes)
                torch.save(data, osp.join(self.processed_dir, 'data_{}.pt'.format(i)))
                i += 1
        print('Processed!')`
so. i need to convert x, edge_index and label to tensor before save.

**Comment by mrqrs** [comment]
`python
 def process(self):

        split = osp.basename(self.root)
        if split == 'train':
            sequences = ['00', '01', '02', '03', '04', '05', '06', '07', '09', '10']
        if split == 'eval':
            sequences = ['08']
        if split == 'test':
            sequences = ['11', '12', '13', '14', '15', '16', '17']
        x_files = get_files_sequence(self.root, sequences, 'voxels', 'npy')
        edge_files = get_files_sequence(self.root, sequences, 'edges', 'npy')
        label_files = get_files_sequence(self.root, sequences, 'labels', 'npy')
        print('processing')
        i = 0
        for j in range(len(x_files)):
            print('Processing:', j, '/', len(x_files))
            voxels_list = np.load(x_files[j], allow_pickle=True)
            edges_list = np.load(edge_files[j], allow_pickle=True)
            labels_list = np.load(label_files[j], allow_pickle=True)
            for m in range(len(voxels_list)):
                x = voxels_list[m]
                num_nodes = x.shape[0]
                edge_index = edges_list[m].T
                label = labels_list[m]
                data = Data(x=x, edge_index=edge_index, y=label, num_nodes=num_nodes)
                torch.save(data, osp.join(self.processed_dir, 'data_{}.pt'.format(i)))
                i += 1
        print('Processed!')
`

so, do you mean that i need to convert the x, edge_index and label to the tensors before save them.?

**Comment by rusty1s** [comment]
Yes, something like `edge_index = torch.from_numpy(edge_index, dtype=torch.long)`

**Comment by mrqrs** [comment]
> Yes, something like `edge_index = torch.from_numpy(edge_index, dtype=torch.long)`

ok. Thank you very much.  👍
